Refactor assorted 'ImageData' SPI

annevk opened the pull request:
#### 3964337dd3210c210631de4d6cea3d9c5a5e49bf
<pre>
Refactor assorted &apos;ImageData&apos; SPI
<a href="https://bugs.webkit.org/show_bug.cgi?id=286796">https://bugs.webkit.org/show_bug.cgi?id=286796</a>
<a href="https://rdar.apple.com/143942946">rdar://143942946</a>

Reviewed by Said Abou-Hallawa.

This aligns WKWebView getInformationFromImageData,
createIconDataFromImageData, and decodeImageData in capabilities.

* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::create):
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::tryCreateNativeImageFromBitmapImageData):
(WebCore::tryCreateNativeImageFromData):
(WebCore::expandNativeImageToData):
(WebCore::expandSVGImageToData):
(WebCore::createIconDataFromImageData):
(WebCore::decodeImageWithSize):
(WebCore::createNativeImageFromData): Deleted.
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::SVGImage):
(WebCore::SVGImage::drawForContainer):
(WebCore::SVGImage::nativeImage):
(WebCore::SVGImage::tryCreateFromData):
(WebCore::SVGImage::isDataDecodable):
* Source/WebCore/svg/graphics/SVGImage.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/SVGImageCasts.cpp:
(TestWebKitAPI::TEST(SVGImageCasts, SVGImageForContainerIsNotSVGImage)):
* Tools/TestWebKitAPI/Tests/WebKit/icon.svg: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageData)):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVG)):

Canonical link: <a href="https://commits.webkit.org/290273@main">https://commits.webkit.org/290273@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4650581501772d7f7cc3dfff07e621996337a286

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44396 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40305 "Failed to checkout and rebase branch from PR 39808") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91590 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17344 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/94530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/40305 "Failed to checkout and rebase branch from PR 39808") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92539 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/6989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39411 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16720 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/96358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16975 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/96358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/21589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9886 "Failed to checkout and rebase branch from PR 39808") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14034 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16733 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16474 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19925 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18256 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -188,6 +190,37 @@ static RefPtr<NativeImage> createNativeImageFromData(std::span<const uint8_t> da
+        if (!newImage)
+            return nullptr;
+
+        CGImageDestinationAddImage(destination.get(), newImage.get(), nullptr);
```
I think this code is a little inefficient. `SVGImageForContainer::nativeImage()` creates an `ImageBuffer` draws the SVGImage to it and then sinks the `ImageBuffer` into a `NativeImage`. This NativeImage is then drawn to another `CGBitmapContext` and then we get a CGImage from `CGBitmapContext`. 

Can't we use something like this:
```
    auto size = FloatSize { length, length };
    RefPtr buffer = ImageBuffer::create(size, RenderingMode::Unaccelerated, RenderingPurpose::Unspecified, 1, 1, DestinationColorSpace::SRGB(), ImageBufferPixelFormat::BGRA8);
    if (!buffer)
        return nullptr;

    Ref svgImageContainer = SVGImageForContainer::create(&svgImage, size, 1, { });
    svgImageContainer->draw(buffer->context(), size, size, { });

    RefPtr newImage = ImageBuffer::sinkIntoNativeImage(WTFMove(buffer));
    if (!newImage)
        return nullptr;

    CGImageDestinationAddImage(destination.get(), newImage.platformImage().get(), nullptr);
```
Have you done this way because you want to pass `kCGImageAlphaPremultipliedLast` to `CGBitmapContextCreate()`?

annevk replied:
I have made your approach work (required a couple minor tweaks). I had essentially copied `expandSVGImageToData` and made minimal changes to get SVG to work. Unfortunately I don't have a deep understanding of all the APIs involved.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -217,19 +250,30 @@ static RefPtr<SharedBuffer> expandNativeImageToData(NativeImage& image, ASCIILit
 {
+    // Supported ICO image sizes by ImageIO.
+    constexpr std::array<unsigned, 5> availableLengths { { 16, 32, 48, 128, 256 } };
+    constexpr auto ico = "com.microsoft.ico"_s;
```
ico -> icoUTI? Or ico -> icoImageType?

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -217,19 +250,30 @@ static RefPtr<SharedBuffer> expandNativeImageToData(NativeImage& image, ASCIILit
+    if (RefPtr svgImage = SVGImage::createFromData(data))
+        nativeImage = svgImage->nativeImage(svgImage->size());
+    else
+        nativeImage = createNativeImageFromData(data, preferredSize);
```
I would reverse the order here. Trying to decode a malformed `BitmapImage` data should return faster than trying to build an `SVGImage` from a malformed `SVGImage` data . This will require creating `Parser`, `Document`, `Page`, `ResourceLoader`, etc.

shallawa replied:
`createNativeImageFromData()` has become a confusing name with this change. 
What does it do exactly?  Does it create a `NativeImage` from any data? I do not think so.

Should it be renamed `tryCreateNativeImageFromBitmapImageData()`?
Should we add a similar one for the `SVGImage`: `tryCreateNativeImageFromSVGImageData()`?

I think adding `try` as a prefix in the name is important since we do not know whether the data is `BitmapImage` or `SVGImage` data or even something else.

shallawa commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -528,16 +536,14 @@ bool isInSVGImage(const Element* element)
 }
 
-bool SVGImage::isDataDecodable(const Settings& settings, std::span<const uint8_t> data)
+RefPtr<SVGImage> SVGImage::createFromData(std::span<const uint8_t> data)
```
I think we should have both functions: `isDataDecodable()` and `createFromData()`. The caller of `isDataDecodable()` does not need to create an `SVGImage` and sets the data to it to know the value of `isDataDecodable()`. The document result of the parser should be enough.

Maybe you can create a function to have a the common code:
```
static RefPtr<Document> createDocumentFromSVGImageData()
{
    ...
}
```

annevk replied:
I left this improvement for the future as I'm not sure how to completely bypass `DocumentLoader` and such.

shallawa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8058,7 +8058,7 @@ void WebPage::getInformationFromImageData(const Vector<uint8_t>& data, Completio
         return completionHandler(makeUnexpected(ImageDecodingError::Internal));
 
-    if (SVGImage::isDataDecodable(m_page->settings(), data.span()))
+    if (SVGImage::createFromData(data.span()))
```
This has become a little inefficient.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -217,19 +248,26 @@ static RefPtr<SharedBuffer> expandNativeImageToData(NativeImage& image, ASCIILit
+    if (!nativeImage) {
+        if (RefPtr svgImage = SVGImage::tryCreateFromData(data))
+            nativeImage = svgImage->nativeImage(svgImage->size());
+    }
```
Can we move this code into a separate function `createNativeImageFromData()` which can make early returns and keep this function untouched:
```
static RefPtr<NativeImage> createNativeImageFromData(...)
{
    if (RefPtr nativeImage = tryCreateNativeImageFromBitmapImageData(data, preferredSize))
        return nativeImage;
    ...
}
```

shallawa commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -185,10 +185,10 @@ ImageDrawResult SVGImage::drawForContainer(GraphicsContext& context, const Float
+    if (observer) {
+        // Temporarily reset image observer, we don't want to receive any changeInRect() calls due to this relayout.
+        setImageObserver(nullptr);
+    }
```
I am not sure about this change. Will it be a big difference between:

```
setImageObserver(nullptr);
...
setImageObserver(WTFMove(observer));
```
versus
```
if (observer)
    setImageObserver(nullptr);
...
if (observer)
    setImageObserver(WTFMove(observer));
```
I personally prefer the first option (do not check for observer) because the case when (observer == nullptr) is used only in this PR and it is very rare case. The first option should work even if observer is nullptr.

shallawa commented on Source/WebCore/svg/graphics/SVGImage.cpp:
```diff
@@ -220,17 +226,19 @@ RefPtr<NativeImage> SVGImage::nativeImage(const DestinationColorSpace& colorSpac
-    setImageObserver(nullptr);
-    setContainerSize(size());
+    if (observer)
+        setImageObserver(nullptr);
```
Ditto.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -188,6 +190,35 @@ static RefPtr<NativeImage> createNativeImageFromData(std::span<const uint8_t> da
+
+        Ref svgImageContainer = SVGImageForContainer::create(&svgImage, size, 1, { });
+        auto rect = FloatRect { { }, size };
+        svgImageContainer->draw(buffer->context(), rect, rect, { });
```
Alternatively these two lines can be written:
```
buffer->context().draw(*svgImageContainer, FloatPoint::zero());
```

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/SVGImageCasts.cpp:
```diff
@@ -82,8 +82,8 @@ class TestImageObserver : public ImageObserver {
 {
-    auto imageObserver = TestImageObserver::create();
-    auto svgImage = SVGImage::create(imageObserver);
+    Ref imageObserver = TestImageObserver::create();
```
I think this observer can be deleted now. But I think it is okay to keep it also.

annevk replied:
I decided to keep this to not reduce test coverage for that code path.

annevk committed:
Refactor assorted 'ImageData' SPI
https://bugs.webkit.org/show_bug.cgi?id=286796
rdar://143942946

Reviewed by Said Abou-Hallawa.

This aligns WKWebView getInformationFromImageData,
createIconDataFromImageData, and decodeImageData in capabilities.

* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::Image::create):
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::tryCreateNativeImageFromBitmapImageData):
(WebCore::tryCreateNativeImageFromData):
(WebCore::expandNativeImageToData):
(WebCore::expandSVGImageToData):
(WebCore::createIconDataFromImageData):
(WebCore::decodeImageWithSize):
(WebCore::createNativeImageFromData): Deleted.
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::SVGImage):
(WebCore::SVGImage::drawForContainer):
(WebCore::SVGImage::nativeImage):
(WebCore::SVGImage::tryCreateFromData):
(WebCore::SVGImage::isDataDecodable):
* Source/WebCore/svg/graphics/SVGImage.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/SVGImageCasts.cpp:
(TestWebKitAPI::TEST(SVGImageCasts, SVGImageForContainerIsNotSVGImage)):
* Tools/TestWebKitAPI/Tests/WebKit/icon.svg: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageData)):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVG)):

Canonical link: https://commits.webkit.org/290273@main
